src/PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.cpp

```diff
@@ -0,0 +1,101 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "PointwiseFunctions/AnalyticSolutions/GrMhd/SmoothFlow.hpp"
+
+#include <algorithm>
+#include <cmath>
+#include <cstddef>
+
+#include "DataStructures/DataVector.hpp"                   // IWYU pragma: keep
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"  // IWYU pragma: keep
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Parallel/PupStlCpp11.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/MakeWithValue.hpp"
+/// \cond
+namespace grmhd {
+namespace Solutions {
+
+SmoothFlow::SmoothFlow(MeanVelocity::type mean_velocity,
+                       WaveVector::type wavevector,
+                       const Pressure::type pressure,
+                       const AdiabaticExponent::type adiabatic_exponent,
+                       const DensityAmplitude::type density_amplitude)
+    :  // clang-tidy: do not std::move trivial types.
+      mean_velocity_(std::move(mean_velocity)),  // NOLINT
+      // clang-tidy: do not std::move trivial types.
+      wavevector_(std::move(wavevector)),  // NOLINT
+      pressure_(pressure),
+      adiabatic_exponent_(adiabatic_exponent),
+      density_amplitude_(density_amplitude) {}
+
+void SmoothFlow::pup(PUP::er& p) noexcept {
+  p | mean_velocity_;
+  p | wavevector_;
+  p | pressure_;
+  p | adiabatic_exponent_;
+  p | density_amplitude_;
+}
+
+template <typename DataType>
+tuples::TaggedTupleTypelist<SmoothFlow::primitive_variables_t<DataType>>
+SmoothFlow::primitive_variables(const tnsr::I<DataType, 3>& x,
+                                const double t) const noexcept {
+  auto result = make_with_value<
+      tuples::TaggedTupleTypelist<SmoothFlow::primitive_variables_t<DataType>>>(
+      x, 0.0);
+
+  // Compute the phase: k * (x - vt)
+  const DataType k_dot_x_minus_vt = [&x, &t, this ]() noexcept {
+    auto local_result = make_with_value<DataType>(x, 0.0);
+    for (size_t i = 0; i < 3; ++i) {
+      local_result += gsl::at(wavevector_, i) *
+                      (x.get(i) - make_with_value<DataType>(
+                                      x, t * gsl::at(mean_velocity_, i)));
+    }
+    return local_result;
+  }
+  ();
+
+  const DataType rest_mass_density =
+      1.0 + density_amplitude_ * sin(k_dot_x_minus_vt);
+  const DataType specific_internal_energy =
+      pressure_ / (adiabatic_exponent_ - 1.0) / rest_mass_density;
+  get<hydro::Tags::RestMassDensity<DataType>>(result) =
+      Scalar<DataType>{rest_mass_density};
+
+  auto velocity = make_with_value<tnsr::I<DataType, 3>>(x, 0.0);
+  for (size_t i = 0; i < 3; ++i) {
+    velocity.get(i) = gsl::at(mean_velocity_, i);
+  }
+  get<hydro::Tags::SpatialVelocity<DataType, 3, Frame::Inertial>>(result) =
+      std::move(velocity);
+
+  get<hydro::Tags::SpecificInternalEnergy<DataType>>(result) =
```

Comment:
Don't introduce `specific_internal_energy`, just compute in place after computing `rest_mass_density` but before moving `rest_mass_density`